rls/src/test/java/io/grpc/rls/internal/AdaptiveThrottlerTest.java

```diff
@@ -0,0 +1,119 @@
+    // initial states
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(0L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(0L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis())).isWithin(TOLERANCE).of(0.0f);
+
+    // Request 1, allowed by all.
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    ticker.advance(1L);
+    throttler.registerBackendResponse(false);
+
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(1L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(0L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis())).isWithin(TOLERANCE).of(0.0f);
+
+    // Request 2, throttled by backend
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    ticker.advance(1L);
+    throttler.registerBackendResponse(true);
+
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(2L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(1L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis()))
+        .isWithin(TOLERANCE)
+        .of(1.0f / 3.0f);
+
+    // Skip half a second (half the duration).
+    ticker.setNow(500L);
+
+    // Request 3, throttled by backend
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    ticker.advance(1L);
+    throttler.registerBackendResponse(true);
+
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(3L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis()))
+        .isWithin(TOLERANCE)
+        .of(2.0f / 4.0f);
+
+    // Request 4, throttled by client.
+    assertThat(throttler.shouldThrottle(0.4f)).isTrue();
+    ticker.advance(1L);
+
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(4L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(3L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis()))
+        .isWithin(TOLERANCE)
+        .of(3.0f / 5.0f);
+
+    // Skip to the point where only requests 3 and 4 are visible.
+    ticker.setNow(1250L);
+
+    assertThat(throttler.requestStat.get(ticker.nowInMillis())).isEqualTo(2L);
+    assertThat(throttler.throttledStat.get(ticker.nowInMillis())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(ticker.nowInMillis()))
+        .isWithin(TOLERANCE)
+        .of(2.0f / 3.0f);
+  }
+
+  private static final class FakeTicker implements Ticker {
+    private AtomicLong currentTimeInMillis = new AtomicLong();
+
+    FakeTicker() {}
+
+    @Override
+    public long nowInMillis() {
+      return currentTimeInMillis.get();
+    }
+
+    long advance(long millis) {
+      return currentTimeInMillis.addAndGet(millis);
+    }
+
+    void setNow(long nowInMillis) {
+      currentTimeInMillis.set(nowInMillis);
+    }
+  }
+}
```

Comment:
Append a newline

Reply:
um, actually i believe internal linter prefers no new line. but i am not sure i should remove from others. lol